File: guests/go/pickers/pickers.go
// The pickers scene (tools/scenes/pickers.steps; docs/datetime-plan.md).
package pickers

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

//go:generate go run dev.kaya/cmd/kaya-gen -type Task -key string
type Task struct {
	Name string
	Due  kaya.Date
}

func App() *kaya.App {
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		dateText := tx.Signal("date: none")
		timeText := tx.Signal("time: none")
		rowText := tx.Signal("row: none")
		dateSig := tx.Signal(kaya.Date{Year: 2026, Month: 9, Day: 4})
		timeSig := tx.Signal(kaya.Time{Hour: 14, Minute: 30})
		tasks := TaskCollection(tx)

		tx.Mount(tx.Column(func() {
			tx.Label(dateText) // label#0
			tx.Label(timeText) // label#1
			tx.Label(rowText)  // label#2
			tx.DatePickerBound(dateSig, func(tx *kaya.Tx, picked kaya.Date) {
				tx.Write(dateText, fmt.Sprintf("date: %v", picked))
			}).
				MinDate(kaya.Date{Year: 2026, Month: 1, Day: 1}).
				MaxDate(kaya.Date{Year: 2026, Month: 12, Day: 31}).
				A11yID("when").A11yLabel("Due") // date_picker#0
			tx.TimePickerBound(timeSig, func(tx *kaya.Tx, picked kaya.Time) {
				tx.Write(timeText, fmt.Sprintf("time: %v", picked))
			}).MinuteStep(15).A11yID("at").A11yLabel("At") // time_picker#0
			tx.Button("reset", func(tx *kaya.Tx) {
				tx.Write(dateSig, kaya.Date{Year: 2026, Month: 3, Day: 1})
				tx.Write(timeSig, kaya.Time{Hour: 9, Minute: 0})
			}) // button#0
			for row := range TaskRows(tx, tasks).All() {
				row.Label(row.Name())
				picker := row.DatePicker(row.Due(),
					func(tx *kaya.Tx, key string, picked kaya.Date) {
						tx.Write(rowText, fmt.Sprintf("row %s: %v", key, picked))
					})
				row.SetA11yID(picker, "due")
			}
		}))

		tasks.Insert(tx, "a", Task{Name: "a", Due: kaya.Date{Year: 2026, Month: 10, Day: 1}})
		tasks.Insert(tx, "b", Task{Name: "b", Due: kaya.Date{Year: 2026, Month: 11, Day: 20}})
	})

	return app
}

File: guests/java/dev/kaya/guests/Pickers.java
package dev.kaya.guests;

import dev.kaya.KayaApp;
import dev.kaya.KayaGen;

import java.time.LocalDate;
import java.time.LocalTime;

/**
 * The pickers scene from the JVM — guests/rust/pickers.rs,
 * tools/scenes/pickers.steps, docs/datetime-plan.md.
 */
public final class Pickers {
    @KayaGen(key = "String")
    record Task(String name, LocalDate due) {}

    private static String day(LocalDate d) {
        return String.format("%04d-%02d-%02d", d.getYear(), d.getMonthValue(),
                d.getDayOfMonth());
    }

    private static String clock(LocalTime t) {
        return String.format("%02d:%02d", t.getHour(), t.getMinute());
    }

    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            KayaApp.Signal<String> dateText = tx.signal("date: none");
            KayaApp.Signal<String> timeText = tx.signal("time: none");
            KayaApp.Signal<String> rowText = tx.signal("row: none");
            KayaApp.Signal<LocalDate> dateSig = tx.signal(LocalDate.of(2026, 9, 4));
            KayaApp.Signal<LocalTime> timeSig = tx.signal(LocalTime.of(14, 30));
            var tasks = TaskKaya.collection(tx);

            tx.mount(tx.column(() -> {
                tx.label(dateText); // label#0
                tx.label(timeText); // label#1
                tx.label(rowText); // label#2
                tx.datePicker(dateSig, LocalDate.of(2026, 1, 1), LocalDate.of(2026, 12, 31),
                                (t, picked) -> t.write(dateText, "date: " + day(picked)))
                        .a11yId("when").a11yLabel("Due"); // date_picker#0
                tx.timePicker(timeSig, 15,
                                (t, picked) -> t.write(timeText, "time: " + clock(picked)))
                        .a11yId("at").a11yLabel("At"); // time_picker#0
                tx.button("reset", t -> { // button#0
                    t.write(dateSig, LocalDate.of(2026, 3, 1));
                    t.write(timeSig, LocalTime.of(9, 0));
                });
                for (var row : TaskKaya.rows(tx, tasks)) {
                    row.label(row.name);
                    var picker = row.datePicker(row.due, (t, key, picked) ->
                            t.write(rowText, "row " + key + ": " + day(picked)));
                    row.setA11yId(picker, "due");
                }
            }));

            tasks.insert(tx, "a", new Task("a", LocalDate.of(2026, 10, 1)));
            tasks.insert(tx, "b", new Task("b", LocalDate.of(2026, 11, 20)));
            return null;
        });

        app.dispatchLoop();
    }

    private Pickers() {}
}
